cli/src/program.rs

```diff
@@ -571,6 +613,55 @@ pub fn parse_program_subcommand(
                 signers: signer_info.signers,
             }
         }
+        ("upgrade", Some(matches)) => {
+            let sign_only = matches.is_present(SIGN_ONLY_ARG.name);
+            let dump_transaction_message = matches.is_present(DUMP_TRANSACTION_MESSAGE.name);
+            let blockhash_query = BlockhashQuery::new_from_matches(matches);
+
+            let Ok(Some(buffer_pubkey)) = pubkey_of_signer(matches, "buffer", wallet_manager)
+            else {
+                return Err(CliError::BadParameter(
+                    "`BUFFER_PUBKEY` must be specified when doing program upgrade".into(),
+                ));
+            };
+
+            let Ok(Some(program_pubkey)) = pubkey_of_signer(matches, "program_id", wallet_manager)
+            else {
+                return Err(CliError::BadParameter(
+                    "`PROGRAM_ID` must be specified when doing program upgrade".into(),
+                ));
+            };
+
+            let (fee_payer, fee_payer_pubkey) =
+                signer_of(matches, FEE_PAYER_ARG.name, wallet_manager)?;
+
+            let mut bulk_signers = vec![
+                Some(default_signer.signer_from_path(matches, wallet_manager)?),
```

Comment:
Oh yeah, I think `generate_unique_signers` will put in the default signer if necessary (if argument wasn't provided by the user),

just forgot about it since it's in "another function" heh :) Fixed in https://github.com/solana-labs/solana/pull/33860/commits/621297fba77f51494b89621486b75f558a39fe42